### ISSUE TITLE
Crash unplugging USB on Windows

### ISSUE THREAD

**Issue by laturetab** [musescore/MuseScore#18161]
### Issue type

Crash or freeze

### Bug description

I use a Scarlett USB on Windows for a microphone and also headphones. If I start Musescore with it plugged in (or not) all is good. If I unplug the USB when Musescore is running it causes an access violation and Musescore crashes (every time.) It does not crash if I plug in the USB when the program is running.

### Steps to reproduce

On Windows 11, run Musescore and plugin a sound device using a USB. Now, unplug the USB and crash.

### MuseScore Version

Latest master (june 24) v4.2

### Regression

I don't know

### Operating system

Windows 11

### Additional context

It crashes in audiodeviceslisterner.cpp in method OnDefaultDeviceChanged() line 121:

if (m_previousIdString == new_default_device_id) {

The first string is a winrt::hstring and the second is a pointer which is NULL when you unplug a USB. The winrt::hstring == operator will blow up if "new_default_device_id" is NULL.

**Comment by laturetab** [comment]
I just reproduced this on the current release build for windows (4.0.2). So it happens both in release and debug builds.

**Comment by laturetab** [comment]
I found 3 problems in OnDefaultDeviceChanged():

1.  When you unplug the USB Audio device, "new_default_device_id" will be NULL and that will cause a crash in the "==" operator every time.

2.  m_previousIdString is not initialized. So if you start with the USB plugged in, m_previousIdString will be empty. Thus the "==" check will not make any sense. Assuming you fix bug 1, the test "m_previousIdString == new_default_device_id" will not do anything (i.e. empty == empty) so it will not call any of the notify methods below. In other words, when you unplug the USB it will ignore it.

3. Assuming you fix 1 and 2 above, it will still sometimes crash if the 2 audio devices don't have the same bit rate. On my machine the speakers were 16bit and the USB was 24bit. Once I made them the same, and fixed 1 and 2 above, everything seems to work.

**Comment by zacjansheski** [comment]
I recreated this issue with my Roland Studio Capture
I also wonder if this is similar to https://github.com/musescore/MuseScore/issues/17836 (removing the current/only audio output device)

[AudioInterfaceCrash.zip](https://github.com/musescore/MuseScore/files/11885725/AudioInterfaceCrash.zip)

**Comment by laturetab** [comment]
Here is the code I added to AudioDevicesListener at the end of the constructor. It fixes (2.) above:

    IMMDevice* defaultDevice = NULL;
    hr = m_deviceEnumerator->GetDefaultAudioEndpoint(eRender, eConsole, &defaultDevice);
    if (FAILED(hr)) {
        LOGE() << "GetDefaultAudioEndpoint failed";
        return;
    }

    LPWSTR id;
    hr = defaultDevice->GetId(&id);
    if (FAILED(hr)) {
        LOGE() << "IMMDevice::GetId failed";
    }
    else {
        m_previousIdString = id;
        CoTaskMemFree(id);
    }
    defaultDevice->Release();

**Comment by laturetab** [comment]
Here is the code I added to OnDefaultDeviceChanged(). It fixes (1.) above.

    // the winrt == operator will blow up if "new_default_device_id" is NULL 
    if (!new_default_device_id) {
        new_default_device_id = L"";
    }

    if (m_previousIdString == new_default_device_id) {
        return S_OK;
    }

**Comment by zacjansheski** [comment]
@laturetab Could you check this PR build and let us know if your issue persists?
https://github.com/musescore/MuseScore/actions/runs/5461260080

I continue to get a crash but I have different devices

**Comment by laturetab** [comment]
The check for NULL works and the program continues and then crashes in a different place: configureDeviceInternal() in function getSharedModeEnginePeriod(). I thought this was caused by different bitrates (3. above) but it seems to be something else. After playing with this for a while I couldn't even get musescore to launch until I rebooted Windows.

There is definitely something off but it is beyond my debugging skills.

Maybe start by seeing why the notification methods get called several times when you unplug the USB. There also seems to be a difference when you load and play a score before unplugging. However, nothing is consistent in how and when it crashes.

I can help you debug this if you give me a little direction on what to look for (and where.)
